delta/service/src/main/scala/ch/epfl/bluebrain/nexus/delta/service/plugin/PluginsLoader.scala

```diff
@@ -25,22 +27,73 @@ class PluginsLoader(loaderConfig: PluginLoaderConfig) {
   private val parentClassLoader = this.getClass.getClassLoader
 
   /**
-    * Loads all the available [[PluginDef]] from each of the discovered jar files
+    * Loads all the available [[PluginDef]] from each of the discovered jar files. In order to solve class loading
+    * issues caused by plugin dependencies, the load operation accumulates all independent plugin classloaders and
+    * adds them to a mutable [[PluginsClassLoader]] which is the parent of all plugin class loaders. It traverses all
+    * the plugin jar files and attempts to load them in multiple passes where a new pass is attempted if the previous
+    * one managed to load some classes but not all.
     */
-  def load: IO[PluginError, (PluginsClassLoader, List[PluginDef])] =
-    for {
-      jarFiles                   <- UIO.delay(loaderConfig.directories.flatMap(loadFiles))
-      (pluginsDef, classLoaders) <- IO.traverse(jarFiles)(loadPluginDef).map(_.flatten.sortBy(_._1).unzip)
-      accClassLoader              = new PluginsClassLoader(classLoaders, parentClassLoader)
-    } yield (accClassLoader, pluginsDef)
+  def load: IO[PluginError, (ClassLoader, List[PluginDef])] = {
+    UIO.delay(loaderConfig.directories.flatMap(loadFiles)).flatMap { jarFiles =>
+      // recursively load the jar files, retrying in case of errors if there's at least on plugin loaded per pass
+      // this enables handling of plugin dependencies
+      IO.tailRecM((jarFiles, new PluginsClassLoader(Nil, parentClassLoader), Nil: List[PluginDef])) {
+        case (Nil, cl, plugins)       => IO.pure(Right((cl, plugins)))
+        case (remaining, cl, plugins) =>
+          // attempt to load the remaining files
+          remaining.traverse(file => loadPluginDef(file, cl).attempt.map(v => (file, v))).flatMap { results =>
+            // partition the load results into List(file -> error) and List(plugin def -> plugin class loader)
+            val partitioned =
+              results.foldLeft((Nil: List[(File, PluginError)], Nil: List[(PluginDef, PluginClassLoader)])) {
+                case ((errors, loaded), (f, Left(err)))          => ((f, err) :: errors, loaded)
+                case ((errors, loaded), (_, Right(None)))        => (errors, loaded)
+                case ((errors, loaded), (_, Right(Some(value)))) => (errors, value :: loaded)
+              }
+
+            partitioned match {
+              // everything was loaded, adding each plugin class loader and return all plugin defs
+              case (Nil, loaded)          =>
+                UIO.delay {
+                  loaded.foreach { case (_, pcl) => cl.addPluginClassLoader(pcl) }
+                  Left((Nil, cl, plugins ++ loaded.map(_._1)))
+                }
+              // nothing resolved, pick the first error and return
+              case ((_, error) :: _, Nil) =>
```

Comment:
We could return all errors no ? 
Or at least log them all and group them by plugin or something ?

Reply:
yep